edg/parts/Lcd_Er_Tft1_28_3.py

```diff
@@ -0,0 +1,105 @@
+
+        dio_model = DigitalBidir.from_supply(
+            self.gnd, self.vdd,
+            voltage_limit_abs=(-0.3 * Volt, self.vdd.voltage_limits.upper()),
+            input_threshold_factor=(0.3, 0.7)
+        )
+
+        self.rs = self.Export(self.conn.pins.request('5').adapt_to(DigitalSink()))
+
+        # Control pins
+        self.spi = self.Port(SpiPeripheral.empty())
+        self.cs = self.Export(self.conn.pins.request('6').adapt_to(dio_model))
+        self.connect(self.spi.sck, self.conn.pins.request('7').adapt_to(dio_model))
+        self.connect(self.spi.mosi, self.conn.pins.request('8').adapt_to(dio_model))
+
+        self.miso_nc = self.Block(DigitalBidirNotConnected())
+        self.connect(self.spi.miso, self.miso_nc.port)
+
+        self.rst = self.Export(self.conn.pins.request('9').adapt_to(DigitalSink()))
+
+        # Capacitive Touch Panel (CTP)
+        self.ctp_i2c = self.Port(I2cTarget(DigitalBidir.empty(), addresses=[0x15]),)
+
+        self.ctp_vdd = self.Export(self.conn.pins.request('10').adapt_to(VoltageSink(
+            voltage_limits=(2.7, 3.6) * Volt,
+            current_draw=(5 * uAmp, 2.5 * mAmp)
+        )))
+
+        self.connect(self.gnd, self.conn.pins.request('11').adapt_to(Ground()))
+
+        self.ctp_rst = self.Export(self.conn.pins.request('12').adapt_to(dio_model))
+        self.ctp_int = self.Export(self.conn.pins.request('13').adapt_to(dio_model))
+        self.connect(self.ctp_i2c.sda, self.conn.pins.request('14').adapt_to(dio_model))
+        self.connect(self.ctp_i2c.scl, self.conn.pins.request('15').adapt_to(dio_model))
+
+
+class Er_Tft_128_3(Lcd, Resettable, Block):
+    """GC9A01-based 1.28" 240x240 TFT, with optional capacitive touch panel."""
+
+    def __init__(self) -> None:
+        super().__init__()
+        self.ic = self.Block(Er_Tft_128_3_Device())
+        self.gnd = self.Export(self.ic.gnd, [Common])
+        self.pwr = self.Export(self.ic.vdd, [Power])
+        self.spi = self.Export(self.ic.spi)
+        self.cs = self.Export(self.ic.cs)
+        self.dc = self.Export(self.ic.rs)
+        # touch interface
+        self.ctp_i2c = self.Export(self.ic.ctp_i2c, optional=True, doc='Touch panel interface i2c')
+        self.ctp_rst = self.Export(self.ic.ctp_rst, optional=True, doc='Touch panel interface')
+        self.ctp_int = self.Export(self.ic.ctp_int, optional=True, doc='Touch panel interface')
+
+    def contents(self):
+        super().contents()
+        self.connect(self.reset, self.ic.rst)
+        self.require(self.reset.is_connected())
+
+        # self.lcd = self.Block(Er_Tft_128_3_Outline())  # for ic outline
+
+        self.connect(self.ic.ledk.adapt_to(Ground()), self.gnd)
+        forward_current = (1, 30)*mAmp
+        self.led_res = self.Block(Resistor(
+            resistance=(self.pwr.link().voltage.upper() / forward_current.upper(),
+                        self.pwr.link().voltage.lower() / forward_current.lower())))
```

Comment:
```suggestion
        forward_voltage = 2.9*Volt
        self.led_res = self.Block(Resistor(
            resistance=((self.pwr.link().voltage.upper() - forward_voltage) / forward_current.upper(),
                        (self.pwr.link().voltage.lower() - forward_voltage) / forward_current.lower())))
```

As we learned it turns out that subtracting the forward voltage is pretty important

There also is a way to do this in one expression (instead of splitting the lower and upper bounds) and have tolerancing work correctly though it's a bit more complex. Can discuss further, might be interesting to try.

Reply:
Applied the current suggestions. 

Is there like an array operations like that?

Reply:
Division can be done as a range operation, `resistance = self.pwr.link().voltage / forward_current`. The problem is that forward current is a spec so the tolerance of resistance needs to shrink, whereas range division semantics here have tolerances expanding.

There is `Range.cancel_multiply`, which does the right thing, but it needs to work on concrete values instead of RangeExpr. I'm thinking of maybe adding `RangeExpr.cancel_multiply`, which can then be used here. Still thinking of syntax, maybe `resistance = (1/forward_current).cancel_multiply(self.pwr.link().voltage)`? Possibly needs a better name too?

Reply:
Does that do element-wise multiplication? what does cancel mean here?